riscv/runtime/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+#![no_std]
+#![feature(start)]
+
+use core::arch::asm;
+use core::panic::PanicInfo;
+
+#[panic_handler]
+fn panic(_panic: &PanicInfo<'_>) -> ! {
+    // activate after #176 has been printed
+    // print(format!("panic at line {} in file {}\n", panic.location().line(), panic.location().file()));
+    unsafe {
+        asm!("unimp");
+    }
+    loop {}
+}
+
+#[inline]
+pub fn get_prover_input(index: u32) -> u32 {
+    let mut value: u32;
+    unsafe {
+        asm!("ecall", lateout("a0") value, in("a0") index);
+    }
+    value
+}
+
+extern "Rust" {
+    fn main();
+}
+#[no_mangle]
+#[start]
+pub unsafe extern "C" fn __runtime_start() -> ! {
+    unsafe {
+        main();
+    }
+    loop {}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memcpy(dest: *mut u8, src: *const u8, n: usize) -> *mut u8 {
+    if (src as u32) % 4 == 0 && (dest as u32) % 4 == 0 {
+        memcpy_aligned(dest, src, n);
+    } else {
+        for i in 0..n {
+            *dest.offset(i as isize) = *src.offset(i as isize);
+        }
+    }
+    dest
+}
+
+pub unsafe fn memcpy_aligned(dest: *mut u8, src: *const u8, n: usize) {
+    let mut i: isize = 0;
+    while i + 3 < n as isize {
+        *((dest.offset(i)) as *mut u32) = *((src.offset(i)) as *mut u32);
+        i += 4;
+    }
+    if i < n as isize {
+        let value = *((src.offset(i)) as *mut u32);
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        *dest_value = (*dest_value & !mask) | (value & mask);
```

Comment:
@chriseth I followed the pattern from your function for the others as well, but I believe that there is an endianness problem here: riscv is little endian from what I read, and if I understand correctly, `mask` is designed for a big endian value.